Client/Frontend/Browser/ResourceDownloadManager.swift

```diff
@@ -0,0 +1,65 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import WebKit
+import Data
+import BraveShared
+
+struct DownloadedResourceResponse: Decodable {
+    let statusCode: Int
+    let data: Data?
+    
+    static func from(message: WKScriptMessage) throws -> DownloadedResourceResponse? {
+        let data = try JSONSerialization.data(withJSONObject: message.body, options: .prettyPrinted)
+        return try JSONDecoder().decode(DownloadedResourceResponse.self, from: data)
+    }
+    
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.statusCode = try container.decode(Int.self, forKey: .statusCode)
+        self.data = Data(base64Encoded: try container.decode(String.self, forKey: .base64Data))
```

Comment:
minor-nit: we should probably add a maximum cap on how much data can be downloaded.

Reply:
I can add that cap in Javascript but how would I know the size of the file without downloading it? It will pull the "PDF" or data from the WebKit's internal cache if available instead of downloading.

Reply:
You can use the `Content-Length` header, if the request is from the cache you can check if the status code is 304.

Reply:
The cache doesn't return a 304 ever. It returns a 0 if it fails. If that's the case, my code returns you null because there is no document available. In the case where a document is available to share, it returns a 200 always and the byte-data associated with it.